### PR TITLE
Branch/save

### PR DESCRIPTION
**Pull request by Martin0071**
Added save functionality

### CODE REVIEW

File: src/main/java/org/example/persistence/ormanager/ORManagerImplTest.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package org.example.persistence.ormanager;
```
Comment (penzata) [inline src/main/java/org/example/persistence/ormanager/ORManagerImplTest.java]: It is best all the test files to be in the test directory.

File: src/main/java/org/example/persistence/ormanager/ORManager.java [hunk 2]
```diff
@@ -49,4 +50,6 @@ static ORManager withPropertiesFrom(String filename) throws SQLException {
     static ORManager withDataSource(DataSource dataSource) {
         return new ORManagerImpl(dataSource);
     }
+    Student save(Student student);
```
Comment (penzata) [inline src/main/java/org/example/persistence/ormanager/ORManager.java]: Wouldn't it be better to use generics for the CRUD methods in the interface.

File: src/main/java/org/example/persistence/ormanager/ORManagerImpl.java [hunk 3]
```diff
@@ -28,4 +28,22 @@ static DataSource createDataSource(String url, String user, String password, Map
         System.out.println(connection.isValid(1000));
         return dataSource;
     }
+    @Override
+    public Student save(Student student){
+        Connection connection;
```
Comment (penzata) [inline src/main/java/org/example/persistence/ormanager/ORManagerImpl.java]: We are going to use connection and statement in all of our methods. Maybe consider making them fields.

File: src/main/java/org/example/persistence/ormanager/ORManagerImpl.java [hunk 4]
```diff
@@ -28,4 +28,22 @@ static DataSource createDataSource(String url, String user, String password, Map
...
+        try {
```
Comment (penzata) [inline src/main/java/org/example/persistence/ormanager/ORManagerImpl.java:87]: Maybe try using try-with-resources here for the connection and the statement.

File: src/main/java/org/example/persistence/ormanager/ORManagerImpl.java [hunk 5]
```diff
@@ -28,4 +28,22 @@ static DataSource createDataSource(String url, String user, String password, Map
...
+            System.out.println("table created");
+           PreparedStatement ps =  connection.prepareStatement("INSERT INTO STUDENTS (first_name) values(?)");
+           ps.setString(1, student.getFirstName());
+           int rows = ps.executeUpdate();
+            System.out.println(rows + " rows affected.");
+
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+        return student;
```
Comment (penzata) [inline src/main/java/org/example/persistence/ormanager/ORManagerImpl.java]: As of this, we are returning the same object (with null Id state) that was passed in the method initially without any change.
Consider getting the auto-generated ID for this object from the database, set it, and then return the saved object with the appropriate id.

File: src/main/java/org/example/persistence/ormanager/ORManagerImpl.java [hunk 6]
```diff
@@ -28,4 +28,22 @@ static DataSource createDataSource(String url, String user, String password, Map
         System.out.println(connection.isValid(1000));
         return dataSource;
     }
+    @Override
+    public Student save(Student student){
+        Connection connection;
+        try {
+           connection = dataSource.getConnection();
+            Statement statement = connection.createStatement();
+            statement.execute("CREATE TABLE IF NOT EXISTS STUDENTS (first_name VARCHAR(255)) ;");
```
Comment (penzata) [inline src/main/java/org/example/persistence/ormanager/ORManagerImpl.java]: About this row and two rows below (row 39): it is always considered good practice to make the sql statements (or the most often used ones) constants and pass a constant in the methods, not the whole string itself.

And about the table creation: right now you are creating a table with just one column. Try checking out the table schema file in the resources directory.

File: src/main/java/org/example/persistence/ormanager/ORManagerImpl.java [hunk 7]
```diff
@@ -28,4 +28,22 @@ static DataSource createDataSource(String url, String user, String password, Map
...
+            System.out.println("table created");
+           PreparedStatement ps =  connection.prepareStatement("INSERT INTO STUDENTS (first_name) values(?)");
+           ps.setString(1, student.getFirstName());
+           int rows = ps.executeUpdate();
+            System.out.println(rows + " rows affected.");
```
Comment (penzata) [inline src/main/java/org/example/persistence/ormanager/ORManagerImpl.java]: Please, don't commit any System.out print lines. If you want to see your current result use Logger for debugging.

File: src/main/java/org/example/persistence/ormanager/ORManagerImpl.java [hunk 8]
```diff
@@ -28,4 +28,22 @@ static DataSource createDataSource(String url, String user, String password, Map
...
+
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
```
Comment (penzata) [inline src/main/java/org/example/persistence/ormanager/ORManagerImpl.java]: Consider handling the exception, not throwing a new one.
Here's a good information about handling that you can use: https://docs.oracle.com/javase/tutorial/jdbc/basics/sqlexception.html

File: src/main/java/org/example/persistence/ormanager/ORManagerImpl.java [hunk 9]
```diff
@@ -1,14 +1,14 @@
...
 import javax.sql.DataSource;
-import java.sql.Connection;
-import java.sql.DriverManager;
-import java.sql.SQLException;
+import javax.xml.crypto.Data;
+import java.sql.*;
 import java.util.Map;
 
-public class ORManagerImpl implements ORManager{
+public class ORManagerImpl implements org.example.persistence.ormanager.ORManager {
```
Comment (penzata) [inline src/main/java/org/example/persistence/ormanager/ORManagerImpl.java]: Why using the whole name of the ORManager interface?

**Review by penzata — CHANGES_REQUESTED** [review]
Please, consider all the comments.

**Review by veselin-petkov — APPROVED** [review]

**Review by penzata — APPROVED** [review]

**Pull request merged**